jabgui/src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java

```diff
@@ -68,29 +74,135 @@ private AutoSetFileLinksUtil(List<Path> directories, ExternalApplicationsPrefere
+     *         linked file with same name exists, add them
+     *         why `add`: Part A are found mainly by CitationKey, which has strong connection to the entry, so we are
+     *                    confident that we should add them automatically
+     * Step 2. try to auto link each broken linked file with Part B.
+     *         how about `unlinked files left in Part B`: there should be no files left as they are found based on broken
+     *                                                    linked files and are used to fix them. `files left` hints a bug
+     *         one trick: we accumulate Part B after Step 1, so Part A does not affect those broken linked files we use
+     *                    to query Part B
+     */
+    private void doLinkAssociatedFiles(BibEntry entry, BiConsumer<List<LinkedFile>, BibEntry> onAddLinkedFile, LinkFilesResult result) {
+        boolean entryUpdated = false;
+        // Step 1: try matched files based on CitationKey configured by user
+        Map<String, LinkedFile> files = getAssociatedFiles(entry, result, preConfiguredFileFinder);
+        List<LinkedFile> newLinkedFiles = autoLinkBrokenLinkedFiles(entry.getFiles(), files);
+        if (filesChanged(newLinkedFiles, entry.getFiles())) {
+            entryUpdated = true;
+            onAddLinkedFile.accept(newLinkedFiles, entry);
+        }
+        // Add left unlinked files as new linked files
+        Set<String> linkedFileNames = entry.getFiles().stream().map(LinkedFile::getLink)
+                                           .map(FileUtil::getBaseName).collect(Collectors.toSet());
+        List<LinkedFile> filesToAdd = files.entrySet().stream()
+                                           .filter(mapEntry -> !linkedFileNames.contains(mapEntry.getKey()))
+                                           .map(Map.Entry::getValue)
+                                           .toList();
+        if (!filesToAdd.isEmpty()) {
+            entryUpdated = true;
+            newLinkedFiles = new ArrayList<>(entry.getFiles());
+            newLinkedFiles.addAll(filesToAdd);
+            onAddLinkedFile.accept(newLinkedFiles, entry);
+        }
+
+        // Step 2: try matched files based on broken linked file names
+        files = getAssociatedFiles(entry, result, brokenLinkedFileNameBasedFileFinder);
+        newLinkedFiles = autoLinkBrokenLinkedFiles(entry.getFiles(), files);
+        if (filesChanged(newLinkedFiles, entry.getFiles())) {
+            entryUpdated = true;
+            onAddLinkedFile.accept(newLinkedFiles, entry);
+        }
+
+        // It is a bug if there are files left
+        if (!files.isEmpty()) {
+            LOGGER.error("Cannot auto-link all the files found based on broken linked file names. Files left: {} ",
+                    files.keySet().stream().map(files::get).map(LinkedFile::getLink).collect(Collectors.joining("||")));
```

Comment:
Strange format, use 

    ["{filename}", "{filename}", "{filename}"]

Easily can be achived with joining.